src/Corax/Querying/Matches/MultiUnaryMatch.cs

```diff
@@ -382,115 +394,145 @@ public static bool Compare(ReadOnlySpan<byte> sx, ReadOnlySpan<byte> sy)
-        Span<bool> allAccepted = _comparers.Length > 128 ? new bool[_comparers.Length] : stackalloc bool[_comparers.Length];
-        allAccepted.Fill(true);
         
-        for (int i = 0; i < read; ++i)
+        do
         {
-            comparerMatches.Fill(false);
-
-            int comparerId = 0;
-            for (; comparerId < _comparers.Length; ++comparerId)
+            int currentMatchesIdx = 0;
+            for (int docIdx = 0; docIdx < read; ++docIdx)
             {
-                ref var comparerMatched = ref Unsafe.Add(ref MemoryMarshal.GetReference(comparerMatches), comparerId);
-                ref var comparer = ref Unsafe.Add(ref MemoryMarshal.GetReference(_comparers.AsSpan()), comparerId);
-                var reader = _searcher.GetEntryTermsReader(matches[i], ref lastPage);
+                var reader = _searcher.GetEntryTermsReader(workingBuffer[docIdx], ref lastPage);
+                var documentMatched = true;
                 
-                while (reader.MoveNext())
+                for (int comparerId = 0; comparerId < _comparers.Length; ++comparerId)
                 {
-                    if(reader.FieldRootPage != Unsafe.Add(ref MemoryMarshal.GetReference(comparerFieldsRootPages), comparerId))
-                        continue;
-                    if (comparerMatched) 
-                        break;
+                    ref var comparer = ref Unsafe.Add(ref MemoryMarshal.GetReference(_comparers.AsSpan()), comparerId);
+                    ref var currentFieldRootPage = ref Unsafe.Add(ref MemoryMarshal.GetReference(comparerFieldsRootPages), comparerId);
                     
-                    var result = IsAcceptedForIterator(comparer, in reader);
-                    if (result)
+                    // We have two modes of UnaryMatch
+                    // Any: that means if any field's term matches our comparer, we've got a match. In this scenario we look for the first matching item.
+                    // All: all terms have been matched by our comparer. In this case we'll search for the first non-matching element.
+                    var isAccepted = MultiUnaryItem.UnaryMode.All == comparer.Mode;
+                    reader.Reset();
+                    while (reader.FindNext(currentFieldRootPage))
                     {
-                        comparerMatched = true;
-                        break;
-                    }
+                        var cmpResult = comparer.Type switch
+                        {
+                            MultiUnaryItem.DataType.Slice => comparer.CompareLiteral(reader),
+                            MultiUnaryItem.DataType.Long => comparer.CompareNumerical(reader), 
+                            MultiUnaryItem.DataType.Double => comparer.CompareNumerical(reader),
+                            _ => ThrowUnknownComparer(comparer)
+                        };
                         
```

Comment:
Can't we know this before the loop and therefore use a managed function pointer to execute the correct one instead? I think the best bet would be a proxy struct with a templated function call which will ensure the entire while get's executed with specialized and packed code.

Reply:
I think he meant something like this:

```
delegate <in EntryTermsReader it, bool> compareDelegate = comparer.Type switch
{
    MultiUnaryItem.DataType.Slice => comparer.CompareLiteral,
    MultiUnaryItem.DataType.Long => comparer.CompareNumerical,
    MultiUnaryItem.DataType.Double => comparer.CompareNumerical,
    _ => ThrowUnknownComparer(comparer)
};
while (reader.FindNext(currentFieldRootPage))
{
    var cmpResult = ccompareDelegate(reader);
```